src/interval.jl

```diff
@@ -222,15 +222,59 @@ end
 Base.:⊈(a::AbstractInterval{T}, b::AbstractInterval{T}) where T = !issubset(a, b)
 Base.:⊉(a::AbstractInterval{T}, b::AbstractInterval{T}) where T = !issubset(b, a)
 
-# Should probably define union, too. There is power in a union.
-
 function Base.intersect(a::AbstractInterval{T}, b::AbstractInterval{T}) where T
     left = max(LeftEndpoint(a), LeftEndpoint(b))
     right = min(RightEndpoint(a), RightEndpoint(b))
     left > right && return Interval{T}()
     return Interval{T}(left.endpoint, right.endpoint, left.included, right.included)
 end
 
+# There is power in a union.
+"""
+    union(intervals::AbstractVector{<:AbstractInterval})
+
+Flattens a vector of overlapping intervals into a new, smaller vector containing only
+non-overlapping intervals.
+"""
+function Base.union(intervals::AbstractVector{<:AbstractInterval})
+    return union!(copy(intervals))
+end
+
+"""
+    union!(intervals::AbstractVector{<:AbstractInterval})
+
+Flattens a vector of overlapping intervals in-place to be a smaller vector containing only
+non-overlapping intervals.
+"""
+function Base.union!(intervals::AbstractVector{<:AbstractInterval})
+    sort!(intervals)
+
+    i = 2
+    while i <= length(intervals)
+        prev = intervals[i - 1]
+        curr = intervals[i]
+
+        # If the current and previous intervals don't intersect then move along
+        if isempty(intersect(prev, curr))
+            i = i + 1
+
+        # If the two intervals intersect then we absorb the current interval into
+        # the previous one.
+        else
+            left = min(LeftEndpoint(prev), LeftEndpoint(curr))
+            right = max(RightEndpoint(prev), RightEndpoint(curr))
+            intervals[i - 1] = Interval(
+                left.endpoint,
+                right.endpoint,
+                Inclusivity(left.included, right.included)
+            )
+            deleteat!(intervals, i)
+        end
+    end
+
+    return intervals
+end
```

Comment:
2 Suggestions:

1. Rather than calling `length` on every loop, just update a variable `n` when we delete an entry.
2. Turn the Interval merging into a constructor.

Example)
```
function Interval(a::Interval, b::Interval)
    isempty(intersect(a, b)) && throw(ArgumentError(...))

    left = min(LeftEndpoint(prev), LeftEndpoint(curr))
    right = max(RightEndpoint(prev), RightEndpoint(curr))
    
    return Interval(
        left.endpoint,
        right.endpoint,
        Inclusivity(left.included, right.included)
    )
end

function Base.merge!(intervals::AbstractVector{<:AbstractInterval{T}}) where T
    sort!(intervals)
    n = length(intervals)

    i = 2
    while i <= n
        prev = intervals[i - 1]
        curr = intervals[i]

        # If the current and previous intervals don't intersect then move along
        if isempty(intersect(prev, curr))
            i = i + 1

        # If the two intervals intersect then we absorb the current interval into
        # the previous one.
        else
            intervals[i - 1] = Interval(prev, curr)
            n = length(deleteat!(intervals, i))
        end
    end

    return intervals
end
```

Reply:
The constructor can already take 2 `Intervals` so it looks like we're getting `merge` after all

Reply:
Lol, fair enough. I'd also be fine with overloading `+`, as we're already using it to extend the endpoints of an interval.

https://github.com/invenia/Intervals.jl/blob/master/src/interval.jl#L141

We could just have a dispatch for `+(::Interval, ::Interval)`.

Reply:
I don't think `+(::Interval, ::Interval)` makes much sense. Currently addition is used to shift a range on the number/date line. I think this would end up being `union(::Interval, ::Interval)` which would need to return a `Set{Interval}` to handle the case where the intervals are disjoint. For now I thing that is out of the scope of this PR.